stackrox/stackrox
# ROX-11070: Add migration for groups bucket.

## Description

This is the fourth PR in a series of PRs regarding removing the composite key from the groups datastore, allowing users to specify different roles for the same group properties.

For full context, you can navigate to the previous PRs by using the comment created below.

This PR will create a migration for the `groups` bucket, adding a ID to all groups which don't have one associated with them as well as using the ID as the new key within the bucket instead of the composite key.

## Testing Performed
- see CI.

Manual testing:
- create some auth providers (number as you like), create a couple of groups per auth provider:
```bash
curl -X POST https://<endpoint>/v1/groupsbatch -H "Authorization: Bearer <token>" -d ‘{"previous_groups":[{"props":{"authProviderId”:”<auth-provider-id>”},”roleName":"Admin"}],"required_groups":[{"props":{"authProviderId":"<auth-provider-id>","key":"name","value":"someday"},"roleName":"None"},{"props":{"authProviderId":"<auth-provider-id>","key":"name","value":"someway"},"roleName":"Scope Manager"},{"props":{"authProviderId":"<auth-provider-id>","key":"email","value":"someone"},"roleName":"Vulnerability Management Requester"},{"props":{"authProviderId":"<auth-provider-id>","key":"email","value":"somewhen"},"roleName":"Admin"},{"props":{"authProviderId":"<auth-provider-id>","key":"email","value":"something"},"roleName":"Vulnerability Report Creator"},{"props":{"authProviderId":"<auth-provider-id>","key":"groups","value":"somehow"},"roleName":"Vulnerability Management Approver"},{"props":{"authProviderId":"<auth-provider-id>","key":"groups","value":"somewhere"},"roleName":"Sensor Creator"},{"props":{"authProviderId":"<auth-provider-id>","key":"userid","value":"someday"},"roleName":"None"},{"props":{"authProviderId":"<auth-provider-id>","key":"userid","value":"sometime"},"roleName":"Continuous Integration"},{"props":{"authProviderId":"<auth-provider-id>","key":"userid","value":"somewhat"},"roleName":"Analyst"},{"props":{"authProviderId":"<auth-provider-id>"},"roleName":"Admin"}]}’
```
- Ensure groups are added correctly within UI as well as returned via API:
```bash
curl https://dh0701wildlimittype.demo.stackrox.com/v1/groups  -H "Authorization: Bearer <token>" 
```
- Start the new tag of the image, ensure that the log indicates a successful migration + contains the sequence number `106`:
```log
pkg/migrations: 2022/07/01 02:59:55.857584 migration_version.go:54: Info: Migration version of database at /var/lib/stackrox/.db-64328521-bef9-4cbe-a584-87de0c203566: &{/var/lib/stackrox/.db-64328521-bef9-4cbe-a584-87de0c203566 3.68.1 92}
Migrator: 2022/07/01 02:59:55.857777 log.go:13: Info: starting DB compaction
Migrator: 2022/07/01 02:59:55.858396 log.go:18: Info: Free fraction of 0.0938 (24576/262144) is < 0.7500. Will not compact
Migrator: 2022/07/01 02:59:56.330146 log.go:18: Info: Found DB at version 92, which is less than what we expect (106). Running migrations...
Migrator: 2022/07/01 02:59:56.336617 log.go:18: Info: Removed 0 service accounts that don't belong to a valid cluster
Migrator: 2022/07/01 02:59:56.349655 log.go:18: Info: Removed 0 K8S roles that don't belong to a valid cluster
Migrator: 2022/07/01 02:59:56.356511 log.go:18: Info: Removed 0 K8S role bindings that don't belong to a valid cluster
Migrator: 2022/07/01 02:59:56.370269 log.go:18: Info: Successfully updated DB from version 92 to 93
Migrator: 2022/07/01 02:59:56.373519 log.go:18: Info: Successfully updated DB from version 93 to 94
Migrator: 2022/07/01 02:59:56.375939 log.go:18: Info: Successfully updated DB from version 94 to 95
Migrator: 2022/07/01 02:59:56.413153 log.go:18: Info: Successfully updated DB from version 95 to 96
Migrator: 2022/07/01 02:59:56.416439 log.go:18: Info: Successfully updated DB from version 96 to 97
Migrator: 2022/07/01 02:59:56.430744 log.go:18: Info: Successfully updated DB from version 97 to 98
Migrator: 2022/07/01 02:59:56.435394 log.go:18: Info: Successfully updated DB from version 98 to 99
Migrator: 2022/07/01 02:59:56.441711 log.go:18: Info: Successfully updated DB from version 99 to 100
Migrator: 2022/07/01 02:59:56.445257 log.go:18: Info: Successfully updated DB from version 100 to 101
Migrator: 2022/07/01 02:59:56.449060 log.go:18: Info: Successfully updated DB from version 101 to 102
Migrator: 2022/07/01 02:59:56.452615 log.go:18: Info: Successfully updated DB from version 102 to 103
Migrator: 2022/07/01 02:59:56.454663 log.go:18: Info: no policy exists for ID 38bf79e7-48bf-4ab1-b72f-38e8ad8b4ec3 in policy migration. Continuing
Migrator: 2022/07/01 02:59:56.458572 log.go:18: Info: Successfully updated DB from version 103 to 104
Migrator: 2022/07/01 02:59:56.461076 log.go:18: Info: Successfully updated DB from version 104 to 105
Migrator: 2022/07/01 02:59:56.474212 log.go:18: Info: Successfully updated DB from version 105 to 106
replica: 2022/07/01 02:59:56.475100 db_replica_manager.go:249: Info: Persisting upgraded replica: temp
version: 2022/07/01 02:59:57.043526 ensure.go:49: Info: Version found in the DB was current. We're good to go!
pkg/migrations: 2022/07/01 02:59:57.044053 migration_version.go:54: Info: Migration version of database at /var/lib/stackrox/current: &{/var/lib/stackrox/current 3.68.1 92}
```
- After the migration, list the groups and verify all of the groups have now an ID associated with them.
- Check within the UI that the groups are correctly listed within each auth provider, deletion and adding of new groups should work + no groups should be lost.
